docs/guide/data_migration.md

```diff
@@ -160,3 +307,313 @@ bin/console ibexa:migrations:kaliop:bulk-convert --recursive --input-directory=k
+        -
+            fieldDefIdentifier: name
+            languageCode: eng-US
+            value: Media
+        # - ...
+    actions: {  }
+    references:
+        -
+            name: ref__content__folder__media
+            type: content_id
+        -
+            name: ref_location__folder__media
+            type: location_id
+        -
+            name: ref_path__folder__media
+            type: path
+
+-
+    type: role
+    mode: create
+    metadata:
+        identifier: foo
+    policies:
+        -
+            module: content
+            function: 'read'
+            limitations:
+                -
+                    identifier: Subtree
+                    values: ['reference:ref_path__folder__media']
+
+```
+
+By default, reference files are located in a separate directory `src/Migrations/Ibexa/references`
+(see [configuration reference](#configuration-reference)
+`ibexa_migrations.migration_directory` and `ibexa_migrations.references_files_subdir` options).
+
+Reference files are **NOT** loaded by default. A separate step (type: "reference", mode: "load", with filename as "value")
+is required. Similarly, saving a reference file is done using type: "reference", mode: "save" step, with filename.
+
+For example:
+```yaml
+-
+    type: reference
+    mode: load
+    filename: 'references.yaml'
+
+-
+    type: reference
+    mode: save
+    # You can also use 'references.yaml', in this case it will be overriden
+    filename: 'new_references.yaml'
+```
+
+!!! note
+
+    You don't need to save references if they are used in the same migration file.
+    References are stored in memory during migration, whether they are used or not.
+
+## How Ibexa Migrations work
+
+!!! note
+
+    The following explain inner workings of migrations.
+    Most users don't need to know this, unless they want to customize migrations.
+
+Ibexa Migrations utilize Symfony Serializer component for converting YAML files into Migration Steps and back.
+A special serializer service (named `"ibexa.migrations.serializer"`) is created, beside the normal `"serializer"`
+one created by the framework (to prevent polluting it with Migration-related services).
+Services tagged with `ibexa.migrations.serializer.normalizer` and `ibexa.migrations.serializer.encoder` tags are attached
+as it's normalizers and encoders respectively (see
+[Symfony serializer docs](https://symfony.com/doc/current/serializer.html#adding-normalizers-and-encoders)).
+
+Steps are converted by services implementing
+`Ibexa\Platform\Bundle\Migration\Serializer\Denormalizer\StepNormalizerInterface` tagged with
+`"ibexa.migrations.serializer.step_normalizer"`.
+
+Once converted to an instance of `Ibexa\Platform\Migration\ValueObject\Step\StepInterface`, they are passed to
+one of `Ibexa\Platform\Migration\StepExecutor\StepExecutorInterface` based services (depending on the type
+of Step), which in turn performs the actual execution of the migration.
+
+## Creating your own Actions
```

Comment:
Note to self -> this could be moved to a separate page in future (this one's getting pretty big)